sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Misc/MiscellaneousTests.cs

```diff
@@ -0,0 +1,245 @@
+            return compilation;
+        }
+
+        private static IEnumerable<MetadataReference> GetReferencesNetCore(string entityFrameworkVersion) =>
+            Enumerable.Empty<MetadataReference>()
+                .Concat(FrameworkMetadataReference.Netstandard)
+                .Concat(NuGetMetadataReference.MicrosoftEntityFrameworkCoreSqlServer(entityFrameworkVersion))
+                .Concat(NuGetMetadataReference.MicrosoftEntityFrameworkCoreRelational(entityFrameworkVersion));
+
+        private Type[] GetSonarAnalyzerTypes(Assembly assembly)
+        {
+            var allAnalyzerTypes = assembly.GetTypes()
+                .Where(IsSonarAnalyzer).ToArray();
+
+            Log($"Analyzer count: {allAnalyzerTypes.Length}");
+
+            return allAnalyzerTypes;
+        }
+
+        private bool IsSonarAnalyzer(Type type) =>
+            (!type.IsAbstract && IsDerivedFromSonarDiagnosticAnalyzer(type));
+
+        private static readonly Type SonarDiagnosticAnalyzerType = typeof(SonarDiagnosticAnalyzer);
+
+        private bool IsDerivedFromSonarDiagnosticAnalyzer(Type type)
+        {
+            var current = type.BaseType;
+
+            while (current != null)
+            {
+                if (current == SonarDiagnosticAnalyzerType)
+                {
+                    return true;
+                }
+                current = current.BaseType;
+            }
+            return false;
+        }
+
+        private SonarDiagnosticAnalyzer CreateAnalyzer(Type analyzerType) =>
+            Activator.CreateInstance(analyzerType) as SonarDiagnosticAnalyzer;
+
+        private bool ExecuteAnalyzerWithTimeout(Compilation compilation, SonarDiagnosticAnalyzer analyzer, int timeoutMs)
+        {
+            var cancellationSource = new CancellationTokenSource(timeoutMs);
+
+            try
+            {
+                DiagnosticVerifier.GetAllDiagnostics(compilation, new[] { analyzer }, CompilationErrorBehavior.FailTest, cancellationSource.Token);
+            }
+            catch(AggregateException ex)
+            {
+                if (ex.InnerExceptions.Any(x => x is OperationCanceledException))
+                {
+                    return false;
+                }
+                throw ex;
+            }
+            return true;
+        }
+
+        private void AssertNoUnexpectedSlowAnalyzers(IEnumerable<ExecutionResult> executionResults, IEnumerable<string> expectedSlowAnalyzers)
+        {
+            var actualSlowAnalyzers = executionResults.Where(x => x.TimedOut).Select(x => x.AnalyzerType.FullName);
+            AssertNoUnexpectedSlowAnalyzers(actualSlowAnalyzers.Except(expectedSlowAnalyzers, StringComparer.OrdinalIgnoreCase));
+
+            var unexpectedlyFastAnalyzers = expectedSlowAnalyzers.Except(actualSlowAnalyzers, StringComparer.OrdinalIgnoreCase);
+
+            if (!unexpectedlyFastAnalyzers.Any())
```

Comment:
to avoid multiple negation ("unexpected fast" and "! Any()") which are harder to follow, I would swap the logic and use

```
if (unexpectedlyFastAnalyzers.Any()) {
	// ... fail the analysis due to faster than expected
} else {
	// ... Log the slow analyzers
}
```